### ISSUE TITLE
Consider specializing copy_closure() for local to local transfers

### ISSUE THREAD

**Issue by ebkalderon** [ebkalderon/merkle-tree-nix-store-thing#4]
Transferring objects between two stores on the same filesystem currently uses hard-linking and/or file copies, which is extremely quick and efficient. But alas, we need to implement support for remote stores for `copy_closure()` to ever truly be useful. As such, we may be forced to use byte streams in `Source::yield_objects()` ([link](https://github.com/ebkalderon/merkle-tree-nix-store-thing/blob/c6979a93c2a06fc2a1d286c8a64d12cedfbe0bec/src/copy.rs#L49-L56)) and `Destination::insert_objects()` ([link](https://github.com/ebkalderon/merkle-tree-nix-store-thing/blob/c6979a93c2a06fc2a1d286c8a64d12cedfbe0bec/src/copy.rs#L74-L79)), even for local-to-local transfers.

This is definitely not ideal, so it would be nice if we could somehow specialize the transport for local-to-local closure transfers to use direct file copies and hard-linking, and fall back to `Vec<u8>` transfers everywhere else. This _almost_ works on stable Rust, but not quite, as demonstrated by this [Rust playground link](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=685b48af463afde8bc446adcbe8eba4b).

In short, due to lack of `impl` specialization in stable Rust, there's no way to express a default type parameter or associated type that will tell the type system to infer `Vec<u8>` for all cases except for local-to-local copies. Perhaps if this feature gets stabilized upstream, we can revisit this. But for now, [even Nix uses byte streams for local-to-local store copies](https://github.com/NixOS/nix/blob/93a8a005defbe5204f739853403ef11dbc016f33/src/libstore/store-api.cc#L776-L785), and that still performs reasonably well. So ultimately, I am going to shelve this idea for now and keep these notes, just in case.
